src/main/java/org/broadinstitute/hellbender/tools/recalibration/RecalibrationTables.java

```diff
@@ -1,69 +1,108 @@
+
+    //bidirectional mappings
+    private final Map<Covariate, NestedIntegerArray<RecalDatum>> covariateToTable;
+    private final Map<NestedIntegerArray<RecalDatum>, Covariate> tableToCovariate;
+
+
+    public RecalibrationTables(final StandardCovariateList covariates) {
+        this(covariates, covariates.getReadGroupCovariate().maximumKeyValue() + 1);
     }
 
-    public RecalibrationTables(final Covariate[] covariates, final int numReadGroups) {
-        tables = new ArrayList<>(covariates.length);
-        for ( int i = 0; i < covariates.length; i++ )
-            tables.add(i, null); // initialize so we can set below
+    public RecalibrationTables(StandardCovariateList covariates, final int numReadGroups) {
+        this.additionalTables = new ArrayList<>();
+        this.allTables = new ArrayList<>();
+        this.covariateToTable = new LinkedHashMap<>();
+        this.tableToCovariate = new LinkedHashMap<>();
 
-        qualDimension = covariates[TableType.QUALITY_SCORE_TABLE.ordinal()].maximumKeyValue() + 1;
+        this.qualDimension = covariates.getQualityScoreCovariate().maximumKeyValue() + 1;
         this.numReadGroups = numReadGroups;
 
-        tables.set(TableType.READ_GROUP_TABLE.ordinal(),
-                 new NestedIntegerArray<>(numReadGroups, eventDimension));
+        //two special tables
+        this.readGroupTable = new NestedIntegerArray<>(numReadGroups, eventDimension);
+        allTables.add(readGroupTable);
+        covariateToTable.put(covariates.getReadGroupCovariate(), readGroupTable);
+        tableToCovariate.put(readGroupTable, covariates.getReadGroupCovariate());
+
+        this.qualityScoreTable = makeQualityScoreTable();
+        allTables.add(qualityScoreTable);
+        covariateToTable.put(covariates.getQualityScoreCovariate(), qualityScoreTable);
+        tableToCovariate.put(qualityScoreTable, covariates.getQualityScoreCovariate());
+
+        //Non-special tables
+        for (Covariate cov : covariates.getAdditionalCovariates()){
+            final NestedIntegerArray<RecalDatum> table = new NestedIntegerArray<>(numReadGroups, qualDimension, cov.maximumKeyValue() + 1, eventDimension);
+            additionalTables.add(table);
+            allTables.add(table);
+            covariateToTable.put(cov, table);
+            tableToCovariate.put(table, cov);
+        }
+    }
+
+    public NestedIntegerArray<RecalDatum> getTableForCovariate(Covariate cov) {
+        return covariateToTable.get(cov);
+    }
+
+    public Covariate getCovariateForTable(NestedIntegerArray<RecalDatum> table) {
+        return tableToCovariate.get(table);
+    }
+
+    public boolean isReadGroupTable(NestedIntegerArray<RecalDatum> table) {
+        return table.equals(getReadGroupTable());
+    }
 
-        tables.set(TableType.QUALITY_SCORE_TABLE.ordinal(), makeQualityScoreTable());
+    public boolean isQualityScoreTable(NestedIntegerArray<RecalDatum> table) {
+        return table.equals(getQualityScoreTable());
```

Comment:
Here too: use `==` instead of `equals()` if you mean to compare memory addresses.